onnxruntime/core/framework/session_state_initializer.cc

```diff
@@ -181,37 +186,48 @@ common::Status SaveMLValueNameIndexMapping(const GraphViewer& graph_viewer,
   return Status::OK();
 }
 
-common::Status DeserializeTensorProto(const ONNX_NAMESPACE::TensorProto& tensor_proto,
-                                      const OrtAllocatorInfo& alloc_info,
-                                      const ExecutionProviders& exec_providers,
-                                      MLValue& mlvalue, void* preallocated, size_t preallocated_size) {
-  auto alloc_ptr = utils::GetAllocator(exec_providers, alloc_info);
-  if (!alloc_ptr) {
-    return Status(common::ONNXRUNTIME, common::FAIL, "Failed to get allocator for alloc_info: " + alloc_info.ToString());
-  }
-
+// This function won't check if the preallocated buffer(p_data) has enough room.
+common::Status DeserializeTensorProto(const Env& env, const std::basic_string<PATH_CHAR_TYPE>& proto_path,
+                                      const ONNX_NAMESPACE::TensorProto& tensor_proto, const MemBuffer& m,
+                                      const ExecutionProviders& exec_providers, MLValue& mlvalue, OrtDeleter& deleter) {
+  const OrtAllocatorInfo& alloc_info = m.GetAllocInfo();
   if (strcmp(alloc_info.name, CPU) == 0 || alloc_info.mem_type == OrtMemTypeCPUOutput) {
     // deserialize directly to CPU tensor
-    return utils::TensorProtoToMLValue(tensor_proto, alloc_ptr, preallocated, preallocated_size, mlvalue);
+    return utils::TensorProtoToMLValue(env, proto_path.c_str(), tensor_proto, m, mlvalue, deleter);
   }
+  // deserialize and copy
+  // The result tensor won't need a deleter because:
+  // 1. It mustn't a string tensor
+  // 2. The memory is not memory-mapped.
+  deleter.f = nullptr;
+  deleter.param = nullptr;
+  if (tensor_proto.data_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "string tensor is not supported for copying between allocators");
+  }
+
+  size_t len;
+  ORT_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<0>(tensor_proto, &len));
 
   std::unique_ptr<Tensor> p_tensor;
   // deserialize to CPU first for non-CPU allocator, then alloc and copy
-  AllocatorPtr deserialize_alloc_ptr;
-  std::unique_ptr<Tensor> p_deserialize_tensor;
-  deserialize_alloc_ptr = exec_providers.Get(kCpuExecutionProvider)->GetAllocator(0, OrtMemTypeDefault);
-  ORT_RETURN_IF_ERROR(utils::GetTensorFromTensorProto(tensor_proto, &p_deserialize_tensor,
-                                                      deserialize_alloc_ptr));
+  AllocatorPtr deserialize_alloc_ptr = exec_providers.Get(kCpuExecutionProvider)->GetAllocator(0, OrtMemTypeDefault);
+  void* buffer = deserialize_alloc_ptr->Alloc(len);
+  MLValue tmp_mlvalue;
```

Comment:
I think this needs to be wrapped in something to guarantee cleanup like IAllocatorUniquePtr. 
You could return on line 217 or 222 prior to this being freed.